### PR TITLE
add sprint management page and backend/db logics

### LINKED ISSUE #90: Ryhmä voi määrittää sprintin alkupäivämäärän ryhmäkohtaisella sivulla
## Tasks

- [ ] [](url) #105  
- [ ] [](url) #92

### PR DESCRIPTION
**Pull request by miahro**
resolves #90

### CODE REVIEW

File: backend/controllers/sprints.js [hunk 1]
```diff
@@ -52,48 +47,99 @@ function validateSprint({ start_date, end_date, sprint }) {
...
+
+    if (groupSprints.length === 0) {
+      console.error('No sprints found for the group')
+      throw new Error('No sprints found for the group')
+    }
+    const formattedSprints = groupSprints.map(sprint => ({
+      id: sprint.id,
+      start_date: new Date(sprint.start_date).toISOString().slice(0, 10),
+      end_date: new Date(sprint.end_date).toISOString().slice(0, 10),
+      sprint: sprint.sprint,
```
Comment (vejol) [inline backend/controllers/sprints.js:69]: What is this field? If sprint means integer 1, 2, 3 etc., consider better naming such as sprint_no

File: backend/controllers/sprints.js [hunk 2]
```diff
@@ -52,48 +47,99 @@ function validateSprint({ start_date, end_date, sprint }) {
...
+      user_id: studentNumber
+    }))
+
+    return formattedSprints
+  } catch (error) {
+    console.error('Error fetching sprtings:', error)
```
Comment (vejol) [inline backend/controllers/sprints.js:75]: Small typo in error message (sprtings)

File: backend/controllers/sprints.js [hunk 3]
```diff
@@ -2,37 +2,32 @@ const express = require('express')
...
+  try {
+    const latestGroup = await db.Group.findOne({
+      include: [{
+        model: db.User,
+        as: 'students',
+        where: { student_number: studentNumber }
+      }],
+      order: [['createdAt', 'DESC']]
+    })
+    if (!latestGroup) {
```
Comment (vejol) [inline backend/controllers/sprints.js:19]: Is this if statement needed? Sequelize findOne method (may?) return null if no matches. In that case line 23 should throw TypeError. 

This is not a big deal but same lines are repeated at the moment (20, 21 -> 26, 27)

File: backend/controllers/sprints.js (hunk 2 above)
Comment (miahro) [reply backend/controllers/sprints.js]: fixed.

File: backend/controllers/sprints.js (hunk 3 above)
Comment (miahro) [reply backend/controllers/sprints.js:19]: Yes and no depending on point of view: it is correct that line 23 would throw TypeError if latestGroup is null. But imo, it is better practice to raise controlled errors and custom error message when you know what is wrong. This makes debugging easier, and in a sense separates uncontrolled and unexpected bugs from controlled error barriers. Prefer to leave as it is.

Yes, the error message should not be the same, these are different cases. The first inside try is that user doesn't belong to a group. One in catch part is for whatever else could have gone wrong. Changed the latter message.

File: backend/controllers/sprints.js (hunk 1 above)
Comment (miahro) [reply backend/controllers/sprints.js:69]: Left as is due to risky change (w/o automated testing). Issue raised for later change.

**Review by vejol — APPROVED** [review]
Looks good to me!

**Pull request merged**